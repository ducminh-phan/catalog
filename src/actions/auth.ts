import { createAction, createAsyncThunk } from "@reduxjs/toolkit";

import Request from "utils/api";

export interface LoginRequest {
  username: string;
  password: string;
}

export interface LoginResponse {
  access_token: string;
}

export const login = createAsyncThunk("auth/login", (body: LoginRequest) => {
  return Request.post<LoginResponse>("login", body);
});

export interface RegisterRequest {
  username: string;
  password: string;
  email: string;
  name: string;
}

export interface RegisterResponse {
  access_token: string;
}

export const register = createAsyncThunk(
  "auth/register",
  (body: RegisterRequest) =>
    Request.post<RegisterResponse>("registrations", body),
);

export const logout = createAction("auth/logout");
