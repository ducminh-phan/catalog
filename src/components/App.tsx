import "assets/styles/App.css";

import {
  AppBar,
  Box,
  Container,
  Grid,
  Paper,
  Toolbar,
} from "@material-ui/core";
import React, { ReactElement } from "react";

import CategoryList from "./CategoryList";
import Header from "./Header";

function App(): ReactElement {
  return (
    <>
      <AppBar position="static">
        <Toolbar>
          <Header />
        </Toolbar>
      </AppBar>
      <Box p={1}>
        <Container>
          <Grid container>
            <Grid item xs={4}>
              <Paper>
                <Box p={1}>
                  <CategoryList />
                </Box>
              </Paper>
            </Grid>
          </Grid>
        </Container>
      </Box>
    </>
  );
}

export default App;
