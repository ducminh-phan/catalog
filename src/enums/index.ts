export const ITEMS_PER_PAGE = 2;

export const CATEGORIES_PER_PAGE = 2;
