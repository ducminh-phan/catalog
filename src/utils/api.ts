import { Json, snakeToCamel } from "./misc";
import storage from "./storage";

type Method = "get" | "post" | "put";

class Request {
  static request = async <T extends Json>(
    endpoint: string,
    method: Method,
    body?: object,
  ): Promise<T> => {
    const token = storage.getToken();
    const headers: HeadersInit = {
      Accept: "application/json",
    };

    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }

    const config: RequestInit = {
      method,
      headers,
    };

    if (body) {
      config.body = JSON.stringify(body);
      headers["Content-Type"] = "application/json";
    }

    const response = await window.fetch(
      `${process.env.REACT_APP_API_URL}/${endpoint}`,
      config,
    );

    const data = snakeToCamel(await response.json());

    if (response.ok) {
      return data as T;
    }

    return Promise.reject(data);
  };

  static get = <T extends Json>(endpoint: string): Promise<T> =>
    Request.request(endpoint, "get");

  static post = <T extends Json>(endpoint: string, body: object): Promise<T> =>
    Request.request<T>(endpoint, "post", body);

  static put = <T extends Json>(endpoint: string, body: object): Promise<T> =>
    Request.request<T>(endpoint, "put", body);
}

export default Request;
