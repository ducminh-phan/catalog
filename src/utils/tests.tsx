import { StylesProvider } from "@material-ui/styles";
import { render as rtlRender, RenderResult } from "@testing-library/react";
import { Rule, StyleSheet } from "jss";
import React, { ComponentType, ReactElement, ReactNode } from "react";
import { Provider } from "react-redux";

import Notification from "components/Notification";
import { AuthProvider } from "contexts/auth";
import { NotificationProvider } from "contexts/notification";
import store from "store";

import Request from "./api";
import { UserResponse } from "./auth";
import storage from "./storage";

// Custom logic to make Material UI class names consistent in snapshots
const generateClassName = (
  rule: Rule,
  styleSheet?: StyleSheet<string>,
): string => `${styleSheet?.options.classNamePrefix}-${rule.key}`;

const render = (ui: ReactElement, renderOptions = {}): RenderResult => {
  const Wrapper = ({ children }: { children: ReactNode }): ReactElement => {
    return (
      <StylesProvider generateClassName={generateClassName}>
        <Provider store={store}>
          <NotificationProvider>
            <AuthProvider>{children}</AuthProvider>
            <Notification />
          </NotificationProvider>
        </Provider>
      </StylesProvider>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper as ComponentType, ...renderOptions });
};

export const setupUser = (): void => {
  storage.setToken("accessToken");
  expect(storage.getToken()).toBe("accessToken");
  jest.spyOn(Request, "get").mockImplementation(
    (): Promise<UserResponse> =>
      Promise.resolve({
        email: "minhπ",
        username: "minhπ",
        created: "",
        updated: "",
        name: "minhπ",
        id: 0,
      }),
  );
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
