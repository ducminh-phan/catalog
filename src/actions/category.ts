import { createAsyncThunk } from "@reduxjs/toolkit";

import Request from "utils/api";
import * as types from "utils/types";

import type { PaginationParams } from "../utils/types";

type FetchCategoriesResponse = {
  totalCategories: number;
  categories: types.Category[];
};

export const fetchCategories = createAsyncThunk(
  "categories/fetch",
  ({ offset, limit }: PaginationParams) =>
    Request.get<FetchCategoriesResponse>(
      `categories?offset=${offset}&limit=${limit}`,
    ),
);

type AddCategoryRequest = {
  name: string;
  description: string;
};

export const addCategory = createAsyncThunk(
  "categories/add",
  (payload: AddCategoryRequest) =>
    Request.post<types.Category>("categories", payload),
);
