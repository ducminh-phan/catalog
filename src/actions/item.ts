import { createAsyncThunk } from "@reduxjs/toolkit";

import Request from "utils/api";
import * as types from "utils/types";

import type { PaginationParams } from "../utils/types";

type FetchItemsRequest = PaginationParams & {
  categoryId?: string;
};

type FetchItemsResponse = {
  totalItems: number;
  items: types.Item[];
};

export const fetchItems = createAsyncThunk(
  "items/fetch",
  ({ categoryId, offset, limit }: FetchItemsRequest) => {
    return Request.get<FetchItemsResponse>(
      `categories/${categoryId}/items?offset=${offset}&limit=${limit}`,
    );
  },
);

export const addItem = createAsyncThunk("item/add", () => Promise.resolve());
