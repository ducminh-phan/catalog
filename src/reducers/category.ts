import { createReducer } from "@reduxjs/toolkit";
import { normalize } from "normalizr";

import { fetchCategories } from "actions/category";
import * as schemas from "utils/schemas";
import * as types from "utils/types";

interface CategoryState {
  totalCategories: number;
  categories: {
    [key: string]: types.Category;
  };
  ids: string[];
}

const initialState: CategoryState = {
  totalCategories: 0,
  categories: {},
  ids: [],
};

export default createReducer(initialState, (builder) =>
  builder.addCase(fetchCategories.fulfilled, (state, action) => {
    const { payload } = action;
    const normalized = normalize(payload.categories, schemas.categories);

    return {
      totalCategories: payload.totalCategories,
      categories: normalized.entities.categories ?? state.categories,
      ids: normalized.result,
    };
  }),
);
