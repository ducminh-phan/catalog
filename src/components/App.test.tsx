import React from "react";

import { FetchCategoriesResponse } from "actions/category";
import App from "components/App";
import { CATEGORIES_PER_PAGE } from "enums";
import Request from "utils/api";
import { render, wait } from "utils/tests";

it("gets the list of categories", async () => {
  jest
    .spyOn(Request, "get")
    .mockImplementation(
      (): Promise<FetchCategoriesResponse> =>
        Promise.resolve({ totalCategories: 0, categories: [] }),
    );

  expect(Request.get).toHaveBeenCalledTimes(0);

  const { baseElement, getByText } = render(<App />);
  await wait(() => getByText("Home"));

  expect(Request.get).toHaveBeenCalledWith(
    `categories?offset=0&limit=${CATEGORIES_PER_PAGE}`,
  );
  expect(baseElement).toMatchSnapshot();
});

it("shows Home/Login/Register", async () => {
  const { getByText, queryByText } = render(<App />);

  await wait(() => getByText("Home"));
  expect(queryByText("Home")).toBeInTheDocument();
  expect(queryByText("Login")).toBeInTheDocument();
  expect(queryByText("Register")).toBeInTheDocument();
  expect(queryByText("Cancel")).toBeNull();
});
