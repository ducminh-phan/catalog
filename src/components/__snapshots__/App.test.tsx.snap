// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets the list of categories 1`] = `
<body>
  <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <div
            class="MuiBox-root MuiBox-root"
          >
            <a
              aria-current="page"
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
              href="/"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                Home
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Register
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
    <div
      class="MuiBox-root MuiBox-root"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
            >
              <div
                class="MuiBox-root MuiBox-root"
              >
                <ul
                  class="MuiList-root MuiList-padding"
                />
                <ul
                  class="MuiList-root MuiList-padding"
                >
                  <li
                    class="MuiListItem-container"
                  >
                    <div
                      class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
                    >
                      <nav
                        aria-label="pagination navigation"
                        class="MuiPagination-root"
                      >
                        <ul
                          class="MuiPagination-ul"
                        >
                          <li>
                            <button
                              aria-label="Go to previous page"
                              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-disabled MuiButtonBase-disabled"
                              disabled=""
                              tabindex="-1"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiPaginationItem-icon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                                />
                              </svg>
                            </button>
                          </li>
                          <li>
                            <button
                              aria-label="Go to next page"
                              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-disabled MuiButtonBase-disabled"
                              disabled=""
                              tabindex="-1"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiPaginationItem-icon"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                                />
                              </svg>
                            </button>
                          </li>
                        </ul>
                      </nav>
                    </div>
                    <div
                      class="MuiListItemSecondaryAction-root"
                    >
                      <div />
                    </div>
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
